### ISSUE TITLE
Embedded templates: ERB, EEX

### ISSUE THREAD

**Issue by dbernheisel** [nvim-treesitter/nvim-treesitter#1352]
**Language informations**

I'm looking to add support for EEX (embedded Elixir) and ERB (embedded Ruby) templates. The file extensions usually are `html.eex` and `html.erb` to indicate the base language is html; it could be any language however such as `.ex.eex` (Embedded Elixir in Elixir) or `.js.erb` (Embedded Ruby in JavaScript).

I see a parser here: https://github.com/tree-sitter/tree-sitter-embedded-template that's hardcoded for handling the HTML base language, but I don't see that it's included in nvim-treesitter.

I'll be happy to contribute PRs for this if someone could point me in the right direction.

**Comment by vigoux** [comment]
Great, thanks for reporting, we'll try to work on that.

So there's some steps we'll need to take from here, and I have some questions:

1. Are you sure the parser you mentioned supports both your languages?
2. Are those filetypes already recognized by nvim, without tree-sitter?
3. Do you want us to do it, or do you want to do it with our help?

**Comment by dbernheisel** [comment]
1. [tree-sitter-embedded-template](https://github.com/tree-sitter/tree-sitter-embedded-template) supports html.erb, but not html.eex. It doesn't seem like it would be a big lift to add html.eex to it since it's almost identical to erb. I could add injections to their repo to support eex, but unsure how that would end up working here for nvim-treesitter. That said, I'm not positive this is the right tool to use (it looks like it is). But unsure how to inform nvim-treesitter which injection to use and when.

2. with a minimal config (`nvim -u NORC` `NVIM v0.5.0-dev+1358-g192ea01ed`), it seems that erb is detected as filetype `eruby`, and `eelixir` is not detected. I currently use this autocmd to set filetypes: `autocmd BufNewFile,BufRead *.eex,*.leex,*.sface,*.drab setf eelixir`.

3. I'll be happy to contribute myself but need help with the bigger picture between neovim and tree-sitter for injections, and since I don't own tree-sitter-embedded-template I'm not sure if there's a plan to use it for nvim-treesitter.

---

I have a branch here that gets it working, but a problem I encounter is how to apply the injections per filetype. I thought of two solutions

1. In the list of parsers, have the same embedded-template repo be used and name them differently, html_erb and html_eex and html_ejs. However, there is an assumption on the .so file that is generated to match the name of the parser (ie embedded_template).
2. Somehow configure nvim-treesitter to use the same parser (embedded_template), but map the injection to use based on the filetype (eruby, eelixir).

Since I'm personally interested in supporting eex first, I can get it to work with what I have on this branch: https://github.com/nvim-treesitter/nvim-treesitter/compare/master...dbernheisel:db-add-embedded-template?expand=1

![image](https://user-images.githubusercontent.com/643967/120384730-8854ce00-c2f4-11eb-8235-9e142863836d.png)

essentially doing this:

```lua
-- my personal lua config
local parser_config = require('nvim-treesitter.parsers').get_parser_configs()
parser_config.embedded_template = {
  install_info = {
    url = 'https://github.com/tree-sitter/tree-sitter-embedded-template',
    files =  { 'src/parser.c' },
    requires_generate_from_grammar  = true,
  },
  used_by = {'eelixir', 'eex', 'leex'}
}

-- In queries/embedded_template/injections.scm
(content) @html @combined
(code) @elixir @combined

-- An empty queries/embedded_template/highlight.scm
```

---

For others that come across this, in the meantime, changing the autocmd to `... setf html` gives decent results but lacking parsing the embedded elixir.

![image](https://user-images.githubusercontent.com/643967/120378426-68211100-c2ec-11eb-98da-09dd686e39a6.png)

**Comment by connorlay** [comment]
Hey all, I just opened a PR https://github.com/nvim-treesitter/nvim-treesitter/pull/1700 adding my new HEEx grammar. This does not solve this issue entirely, but it does improve support for the use-case of writing HTML in EEx templates.

**Comment by nathanmsmith** [comment]
@vigoux I'm interested in contributing support for .erb files via [tree-sitter-embedded-template](https://github.com/tree-sitter/tree-sitter-embedded-template), since erb is still currently unsupported. Who/where would be the best resource to go to for questions I have while working on a PR?

**Comment by vigoux** [comment]
The best way is always to come to the Zulip server. You should have averything you need in the `CONTRIBUTING.md`.

**Comment by connorlay** [comment]
Here's another PR related to this thread: https://github.com/nvim-treesitter/nvim-treesitter/pull/2351 adds support for EEx and LEEx files by integrating with [tree-sitter-eex](https://github.com/connorlay/tree-sitter-eex).

**Comment by rockerBOO** [comment]
I went down this path as well and I made a [html-eex parser](https://github.com/rockerBOO/tree-sitter-html-eex/) based on the HTML one with some simple capturing. This allows us to do eex injection on top of the html parser. The HTML parser errors when seeing <% and that makes it hard to use to inject.

I use a new filetype, `html_eex`, which is setup using this autocmd.

```lua
vim.api.nvim_create_autocmd({"BufRead", "BufNewFile"}, {
  pattern = "*.html.eex",
  command = "set filetype=html_eex"
})
```

![2022-08-03_02-37](https://user-images.githubusercontent.com/15027/182512287-a50d918e-b8f5-4a4d-84d2-afbfaa5befc8.png)

More info about setting it up on the README. 

Note: I just got it to work, so naming, or other bugs may be there. Any suggestions or improvements would be appreciated as well.

**Comment by mhanberg** [comment]
I am curious if anyone knows if this very patch can make doing this any easier? https://github.com/neovim/neovim/commit/ddd257f75301a50c177fc24a693d39a45b47a689

In general, I am wondering if it is possible to do the following

Have an arbitrary eex or erb file, let's call it `foo.<PRIMARY LANG>.eex`

then the nvim filetype get set to `<PRIMARY LANG>.eex` rather than `eelixir`.

Then have the eex treesitter query say "main syntax is <PRIMARY LANG>, things like `<%= ... %>` is eex, which would thing make the content inside those elixir"

the goal being that the primary lang always dynamic, so you don't have to set up individual filetypes/grammers/etc for each language combo.

At least in Elixir, the _most_ common would be *.html.eex files, but you often also have *.ex.eex files.

hopefully that made some sense.

**Comment by kaka-ruto** [comment]
Would be nice to have this one brought onto TS :)

**Comment by clason** [comment]
PR welcome. Nvim-treesitter is a community effort.

**Comment by kaka-ruto** [comment]
@clason I'd like to! I would need some guidance though, I am new to this. I'd like to add support for Embedded Ruby (ERB). Is it sufficient if I do it like it was done in this PR? https://github.com/nvim-treesitter/nvim-treesitter/pull/2351

**Comment by mhanberg** [comment]
That PR doesn't really address what this issue is about.

**Comment by kaka-ruto** [comment]
> That PR doesn't really address what this issue is about.

I agree it does't. I'd like to add support for erb. Should I open a new issue for that? Re: maintainers.

**Comment by brian-tran-dev** [comment]
I know absolutely nothing about tree sitter works, but I want to use this [tree-sitter-embeded-template repo](https://github.com/tree-sitter/tree-sitter-embedded-template) to support `*.ejs` files. Which config should I set (I use the legacy master branch)? if I need to do something more than just adding some configs, which guides should I follow to use the repo for highlighting and indenting?
